jsonComposer.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2024-Present Couchbase, Inc. All rights reserved.
+
+package gojsonsm
+
+import (
+	"fmt"
+)
+
+var (
+	ErrNilComposer        error = fmt.Errorf("nil composer")
+	ErrNilData            error = fmt.Errorf("nil data")
+	ErrInvalidJSON        error = fmt.Errorf("invalid JSON object")
+	ErrUnexpectedEOF      error = fmt.Errorf("unexpected EOF")
+	ErrInsufficientMemory error = fmt.Errorf("insufficient memory allocated for dst, cannot proceed")
+	ErrUnrecognisedToken  error = fmt.Errorf("unrecognised token in the JSON object")
+)
+
+type jsonObjComposer struct {
+	// should have enough length
+	body []byte
+	// cursor
+	pos int
+	// if the newly composed json has atleast one field
+	atLeastOneFieldLeft bool
+}
+
+// writes a given slice of data to the composer
+// returns any errors in the process
+func (composer *jsonObjComposer) Write(data []byte, tknType tokenType) error {
+	if composer == nil {
+		return ErrNilComposer
+	}
+	if data == nil {
+		return ErrNilData
+	}
+	// if we are about to write objectEnd token i.e. "}", we should ensure that the previous position was not a ","
+	// if it is ",", step back by one position and write a "}"
+	if tknType == tknObjectEnd && composer.pos-1 >= 0 && composer.body[composer.pos-1] == ',' {
```

Comment:
This ',' should probably be using one of the tkn constants.

Reply:
Done